README.md
# lyceum
[![Clojars Project](http://clojars.org/lyceum/latest-version.svg)](http://clojars.org/lyceum)

A Clojure library designed to help in the authoring and testing of modules
riemann rules.

Each rule lives and are distributed in different namespace.
Lyceum will take care to iterate the classpath and look for namespaces
containing rules.

Each namespace can be put under test using a test fixture provided by lyceum
that takes care to setup and tear down all necessary state to simulate riemann
operation.

Tests are run with little overhead, it is not required to start a riemann
core, instead lyceum takes the approach of completely _simulating_ riemann
operation while emposing a little bit of structure on the way you write rules
to make things manageable.

Some key points.

* When running tests, all external interaction is faked and verifiable (using
  `check-externals`).
* The riemann schedules is replaced with a global override of
  `riemann.time/schedule!` that has a thread-local implementation without global
  state.
  With this comes also faked time (`riemann.time/unix-time` et. al.).
* Provides a [restful HTTP service](#http-service) for evaluating rules on-the-fly.

# Getting Started

Start by initializing an empty leiningen project;

```
#> lein new my-rules
#> cd my-rules
#> rm src/my_rules/core.clj
#> rm test/my_rules/core_test.clj
```

Add a testing scope dependency to lyceum in your project.clj

```clojure
(defproject my-rules "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.5.1"]
                 ; add both as a test dependency.
                 [lyceum "0.1.0" :scope "test"]]
  ; ... and as a plugin.
  :plugins [[lyceum "0.1.0"]]
  ; define which namespace lyceum should initialize new rules in.
  :lyceum-namespace my-rules.rules)
```

Download templates and generate a new namespace skeleton.

```
#> mkdir templates
#> wget https://raw.githubusercontent.com/spotify/lyceum/master/templates/rule.clj -O templates/rule.clj
#> wget https://raw.githubusercontent.com/spotify/lyceum/master/templates/test.clj -O templates/test.clj
```

Now it's time to initialize rules for a group called __my-group__.

```
#> lein lyceum init my-group
```

Almost there, verify that your rules are working by running your skeleton
test-cases!

```
#> lein test
```

Now you can start modifying *src/my\_rules/rules/my\_group.clj* to suit your
needs.

Any test-cases you come up with should be added to
*test/my\_rules/rules/my_group_test.clj*, use the generated one as inspiration
for writing more.

When you are done, compile your rules.

```
#> lein uberjar
```

Add lyceum and the rules jar containing to the classpath of riemann and
add the following declaration in your riemann.config.

```clojure
; load-plugins has to be present!
(load-plugins)

(streams
  ; load any namespaces containing rules under 'my-rules.rules'.
  ; Placklist any tutorial namespaces to avoid loading them.
  (lyceum/load-rules
    'my-rules.rules
    :opts {:index index}
    :blacklist [
      #"my-rules.rules.tutorial\d+"
    ]))
```

Make sure that you start the riemann service with the system property
`-Dlyceum.mode=real` (pass this as an argument to _java_ like `java
-Dlyceum.mode=real ... riemann.bin`).
See the [externals section](#externals) for more details about this.

# Time

Time is controlled in a similar fashion to how `riemann.controlled.time` works,
but it's done with a separate implementation in lyceum.

Any input event is inspected for its `:time` field, and if present, the time
specified is used as the current time for the simulation.

This can bee seen in the example test-case that you generated if you followed
the guide above.

# Externals

Lyceum uses __externals__ to interact with external systems, externals are thin
wrappers around the riemann's external integrations (email, pagerduty, etc.).

By default lyceum uses the __fake__ mode which will cause any external
interactions to be logged instead of realized.
This is what allows the test-cases to verify external effects with
`check-externals`.

However when running in production, the mode is set to __real__. This will
cause the wrapping to be discarded and the real external interaction to be
realized.

The mode is set with the `-Dlyceum.mode=<mode>` system property that should
be passed to your JVM.

Valid modes are.

* __fake (default)__ - Log any externals that are triggered to an internal
  data-structure, allowing for later verification.
* __test__ - Write external interaction to a log file.
* __real__ - Realize any externals that are triggered.

Wrapping your own external is straight forward, you can use
[the email external](src/lyceum/external/email.clj) as an example for how to do
this.

# HTTP Service

You can start the HTTP service by running.

```
#> lein run [lyceum.conf]
```

Or the ___lyceum.service___ class in the resulting uberjar.

```
#> java -jar <path-to-jar> lyceum.service [lyceum.conf]
```

It expects to find a [lyceum.conf](lyceum.conf) in the current working directory, or one can be provided as an argument.

The service is currently capable of loading rules the following ways.

+ Github through [github-rules](src/lyceum/service/rules_loader/github.clj)
+ Filesystem through [directory-rules](src/lyceum/service/rules_loader/directory.clj)

#### POST /eval

Will evaluate the received data (___VERY UNSAFE___) and apply the provided rules to them.

+ Response 200 (application/json)
+ Response 500 (application/json)

###### Request Structure
```javascript
{"data": <string>, "events": [<event>, ..]}
```

###### Response Structure
```javascript
{/* contains any external events (like pagerduty) that happened */
 "reports":[<report>, ..],
 /* contains the events which was indexed during this evaluation. */
 "index":[<event>, ..]
}
```

###### Example CURL
```
#> curl http://localhost:8080/eval -H "Content-Type: application/json"
          -d '{"data": "(ns hello.world) (defn rules [{:keys [index]}] (fn [e] (index e)))", "events": [{"service": "foo"}]}'
```

#### GET /ns/

Will request a list of namespaces, see [lyceum.conf](lyceum.conf) for how this is configured.

#### GET /ns/{ns}

Will request the content of a specific namespace ___{ns}___, see [lyceum.conf](lyceum.conf) for how this is configured.


debian/changelog
spotify-lyceum (0.1.1) unstable; urgency=low

  * Placeholder.

 -- Hero Squad <hero@spotify.com>  Wed, 23 Oct 2013 17:40:35 +0200


src/lyceum/service.clj
;; $LICENSE
;; Copyright 2013-2014 Spotify AB. All rights reserved.
;;
;; The contents of this file are licensed under the Apache License, Version 2.0
;; (the "License"); you may not use this file except in compliance with the
;; License. You may obtain a copy of the License at
;;
;;     http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
;; WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
;; License for the specific language governing permissions and limitations under
;; the License.

;; # Experimental http server for testing rules.
;;
;; * Start: __lein run__
;; * Surf to http://localhost:8080
(ns lyceum.service
  (:import
    java.io.InputStreamReader
    java.io.ByteArrayInputStream
    java.io.PushbackReader
    lyceum.HttpException)
  (:require
    [lyceum.service.rules-loader :as loader]
    [lyceum.service.config-file :as config-file])
  (:require
    riemann.logging
    riemann.common
    riemann.time)
  (:require
    [lyceum.test :refer [with-mocked-riemann-time make-send-events-fn]]
    [lyceum.external :refer [*external-reports*]]
    [clojure.tools.logging :refer [info error]]
    [cheshire.core :as json]
    [compojure.route :refer [files not-found]]
    [compojure.handler :refer [site]]
    [compojure.core :refer [routes GET POST DELETE ANY context]]
    [org.httpkit.server :refer [run-server]])
  (:gen-class :name lyceum.service))

(defn handle-json
  [handler & args]
  (let [headers {"Content-Type" "application/json"}
        serialize json/generate-string
        respond (fn [status body]
                  {:status status
                   :headers headers
                   :body (serialize body)})]
    (fn [req]
      (try
        (respond 200 (apply handler (concat [req] args)))
        (catch HttpException e
          (respond (.getStatus e) {:message (.getMessage e)}))
        (catch Exception e
          (error e "An error happened in the request handler")
          (respond 500 {:message (str "An error occured: " e)}))))))

(defn- http-list-rules [req rules-loader]
  {:rules (loader/list-rules rules-loader)})

(defn- http-get-rule [req rules-loader]
  (let [p-ns (-> req :params :ns)
        rule (loader/get-rule rules-loader p-ns)]
    (if (nil? rule)
      (throw (HttpException. 404 (str "No such rule: " p-ns)))
      {:data rule})))

(defn read-expressions
  [reader]
  (binding [*read-eval* false]
    (loop [l []]
      (let [result (read reader false nil)]
        (if (nil? result)
          l
          (recur (conj l result)))))))

(defn- index-add
  [index e]
  (conj index {:time (riemann.time/unix-time) :event e}))

(defn eval-expressions
  [expressions events current-time]
  (let [temp-ns-name (gensym "lyceum-eval-ns___")
        current-ns (create-ns temp-ns-name)
        evaled-tasks (atom [])]
    (try
      (with-mocked-riemann-time evaled-tasks current-time
        (binding [*ns* current-ns]
          (doseq [expr expressions]
            (if (= (first expr) 'ns)
              (eval (cons 'ns (cons temp-ns-name (rest (rest expr)))))
              (eval expr)))))

      (let [reports (atom [])
            index (atom [])
            make-rules-fn (ns-resolve current-ns 'rules)]
        (when-not make-rules-fn
          (throw (Exception. "No 'rules' function in evaluated namespace.")))

        (let [index-fn (fn [e] (swap! index index-add e))
              rules-fn (make-rules-fn {:index index-fn})]
          (when-not (and rules-fn (fn? rules-fn))
            (throw (Exception. "Calling 'rules' did not return a stream function.")))

          (let [send-events (make-send-events-fn rules-fn)]
            (binding [*external-reports* reports]
              (send-events events))
            {:reports @reports :index @index})))

      (finally
        (remove-ns temp-ns-name)))))

(defn- make-pushback-reader
  [data]
  (let [input-stream (ByteArrayInputStream. (.getBytes data))
        input-stream-reader (InputStreamReader. input-stream)]
    (PushbackReader. input-stream-reader) ))

(defn- http-eval-rule [req]
  (let [json-body-string (slurp (:body req))
        json-body (or (json/parse-string json-body-string) {})
        data (json-body "data" "")
        events (json-body "events" [])
        current-time (int (json-body "current-time" 0))]
    (when (or (nil? data) (empty? data))
      (throw (HttpException. 400 "'data' must not be empty")))
    (when (or (nil? events) (empty? events))
      (throw (HttpException. 400 "'events' must not be empty")))
    (let [pushback-reader (make-pushback-reader data)]
      (try
        (let [expressions (read-expressions pushback-reader)]
          (eval-expressions expressions events current-time))
        (catch RuntimeException e
          (error e "Failed to handle request")
          (throw (HttpException. 400 (str "Invalid body: " e))))))))


(defn make-routes
  [c]
  (let [rules-loader (:rules-loader c)]

    (when (nil? (:rules-loader c))
      (throw (Exception. "No rules-loader specified in lyceum configuration")))

    (routes
      (GET "/rules" [] (handle-json http-list-rules rules-loader))
      (GET "/rules/:ns" [] (handle-json http-get-rule rules-loader))
      (POST "/eval" [] (handle-json http-eval-rule))
      (not-found "not found"))))

(defn http-server
  [listen-port site-routes]
  (info (str "Listening on port " listen-port))
  (run-server (site site-routes) {:port listen-port}))

(defn -main
  [& argv]
  (riemann.logging/init)
  (let [config-path (or (first argv) "lyceum.conf")
        c (config-file/setup config-path)
        listen-port (:listen-port c)
        site-routes (make-routes c)]
    (http-server listen-port site-routes)))
